File: app/src/main/java/com/checkin/app/checkin/Auth/DeviceTokenService.java
package com.checkin.app.checkin.Auth;

import android.app.IntentService;
import android.content.Intent;
import android.preference.PreferenceManager;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.Log;

import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Utility.Constants;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.InstanceIdResult;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DeviceTokenService extends IntentService implements OnSuccessListener<InstanceIdResult>, Callback<ObjectNode> {
    private static final String TAG = DeviceTokenService.class.getSimpleName();

    public static final String KEY_TOKEN = "device_token";

    public DeviceTokenService() {
        super(TAG);
        Log.e(TAG, "Service started!!");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if (intent != null && intent.hasExtra(KEY_TOKEN)) {
            saveToken(intent.getStringExtra(KEY_TOKEN));
        } else {
            FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(this);
        }
    }

    @Override
    public void onSuccess(InstanceIdResult instanceIdResult) {
        String token = instanceIdResult.getToken();
        saveToken(token);
    }

    private void saveToken(String token) {
        AuthRepository.getInstance(getApplication())
                .postDeviceToken(token)
                .enqueue(this);
    }

    private void processResponse(ApiResponse<ObjectNode> response) {
        if (response.isSuccessful()) {
            PreferenceManager.getDefaultSharedPreferences(getApplicationContext())
                    .edit()
                    .putBoolean(Constants.SP_SYNC_DEVICE_TOKEN, true)
                    .apply();
        }
    }

    @Override
    public void onResponse(@NonNull Call<ObjectNode> call, @NonNull Response<ObjectNode> response) {
        processResponse(new ApiResponse<>(response));
    }

    @Override
    public void onFailure(@NonNull Call<ObjectNode> call, @NonNull Throwable t) {
        processResponse(new ApiResponse<>(t));
    }
}

File: app/src/main/java/com/checkin/app/checkin/auth/services/DeviceTokenService.kt
package com.checkin.app.checkin.auth.services

import android.app.IntentService
import android.content.Intent
import android.preference.PreferenceManager
import android.util.Log
import com.checkin.app.checkin.auth.AuthRepository
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.utility.Constants
import com.fasterxml.jackson.databind.node.ObjectNode
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.InstanceIdResult
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class DeviceTokenService : IntentService(TAG), OnSuccessListener<InstanceIdResult>, Callback<ObjectNode> {
    override fun onHandleIntent(intent: Intent?) {
        when (intent?.takeIf { it.hasExtra(KEY_TOKEN) }) {
            null -> FirebaseInstanceId.getInstance().instanceId.addOnSuccessListener(this)
            else -> saveToken(intent.getStringExtra(KEY_TOKEN))
        }
    }

    override fun onSuccess(instanceIdResult: InstanceIdResult) = saveToken(instanceIdResult.token)

    private fun saveToken(token: String) {
        AuthRepository.getInstance(application)
                .postDeviceToken(token)
                .enqueue(this)
    }

    private fun processResponse(response: ApiResponse<ObjectNode>) {
        if (response.isSuccessful) {
            PreferenceManager.getDefaultSharedPreferences(applicationContext)
                    .edit()
                    .putBoolean(Constants.SP_SYNC_DEVICE_TOKEN, true)
                    .apply()
        }
    }

    override fun onResponse(call: Call<ObjectNode>, response: Response<ObjectNode>) {
        processResponse(ApiResponse(response))
    }

    override fun onFailure(call: Call<ObjectNode>, t: Throwable) {
        processResponse(ApiResponse(t))
    }

    companion object {
        const val KEY_TOKEN = "device_token"
        private val TAG = DeviceTokenService::class.java.simpleName
    }

    init {
        Log.e(TAG, "Service started!!")
    }
}